Initial draft of MEDIA_SANITIZE protocol with NVM Express Support

rayrobles opened the pull request:
Initial draft of MEDIA_SANITIZE_PROTOCOL and native NVM Express admin command support for FormatNVM and Sanitize.

NOTE: The MEDIA_SANITIZE_PROTOCOL is placed in MdeModulePkg

rayrobles committed:
Initial draft of EFI_MEDIA_SANITIZE protocol and native NVM Express command support for FormatNVM and Sanitize

rayrobles committed:
Initial draft of EFI_MEDIA_SANITIZE protocol and native NVM Express command support for FormatNVM and Sanitize

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+  /// A pointer to the EFI_BLOCK_IO_MEDIA data for this device.
+  /// Type EFI_BLOCK_IO_MEDIA is defined in BlockIo.h.
+  ///
+  EFI_BLOCK_IO_MEDIA           *Media;
```
why is Media needed in the protocol? 
1. The media is already on the block io that i assume is on the same handle.  
2. None of the api calls here use the Media.  I do the MediaId parameter passed in.  Is that to make sure the caller is operating on what they expect?  If so, I still think the caller could get the data from blockio protocol instead.

rayrobles replied:
1. It's actually not... good catch. I was using other block IO protocols as templates and simply missed removing unnecessary fields. I will remove Media from the protocol.

2. The MediaId is used to ensure the caller is operating on the expected Media specified in the block IO protocol. There is a small windows where a format or other destructive operation could occur, and result in a new MediaId. The caller could have a cached copy of the MediaId (or any private data structures). This is to help ensure the media has not changed or been removed.

spbrogan replied:
so you are expecting that the caller locate blockio get the media id....then locate sanitize on the same handle and call with the media id they got from the blockio?

rayrobles replied:
Ultimately, I see the caller passing in a LUN Id or Namespace Id as the Media Id. The caller does not required to have the block io protocol, but may likely have it. 

I understand your point/concern however. If the caller has access to the actual Block IO or Media Sanitize protocol, then there is no need to pass in the Media Id. 

If the assumption can be made that Block IO is installed on every LUN, namespace, etc., then I'm fine removing Media Id as a parameter. The caller can match their tracked LUN, namespace, etc. on their end first.

spbrogan replied:
i wasn't trying to say the mediaId isn't a valid parameter or a good sanity check.  Seems like there is/might be a design where the block io changes content but protocol instance (memory location) stays the same.  In that case it seems reasonable for the caller to get the media id at the beginning when locating this and then use that for future calls to make sure they are still talking to the expected device?  I was just trying to understand the process here.

rayrobles replied:
That was actually the precise original intent :-). Matching of the expected media vs. any changes that may have happened in a different context or boot (i.e. like an LBA format change or hot swap).

rayrobles replied:
I am going to resolve this comment since it appears you understood my original intent. Please let me know if you'd like to go in a different direction. Thanks.

spbrogan commented on MdePkg/MdePkg.dec:
```diff
@@ -1893,6 +1893,9 @@
   gEfiEraseBlockProtocolGuid           = { 0x95a9a93e, 0xa86e, 0x4926, {0xaa, 0xef, 0x99, 0x18, 0xe7, 0x72, 0xd9, 0x87 }}
 
+  ## Include/Protocol/MediaSanitize.h
+  gEfiMediaSanitizeProtocolGuid        = { 0x0d799a99, 0x25af, 0x429e, {0x92, 0x72, 0xd0, 0xb2, 0x7d, 0x6d, 0x5f, 0x14 }}
```
With all the "Efi" usage I am curious your thoughts.  That should only be used for things defined in the UEFI specification.  Are you going to go for "code first" and get spec approval quickly?  Does this need UEFI spec approval to meet the goals?   This same comment goes for adding it to the MdePkg.  MdePkg in general is only allowed for industry specifications.

rayrobles replied:
This is a great question Sean. I actually had the same concern myself on this feature. Ideally, I would like to update the UEFI spec to include this protocol, but not sure of the method within Microsoft. We can talk offline if necessary.

My goal here is to provide a protocol that maps NIST clear and purge requests to NVMe FormatNVM and Sanitize commands. Does Mu support "private" or non-public protocols to be used? If so, then I have no issue moving the protocol to a more appropriate location. 

However, if all protocols must reside and be defined in MdePkg, then I would like to start the process of standardizing the sanitize protocol. Please advise on next steps.

spbrogan replied:
There are lots of protocols defined in the code tree.  
Some are UEFI defined. 
Some are Microsoft defined in specs on what was MSDN
Some are edk2 defined for use within Edk2 based projects
Some are defined by Project Mu
Some are defined by silicon partners for a given silicon implementation.  .......
you get the idea.  

So it just depends on what you business case is and what group you want to work with to get it defined.  Obviously Edk2/Mu have less work.  UEFI isn't that difficult either given you have buttoned up your use case and are willing to spend a few hours for a few weeks on the meetings.  

Since here you are adding it to mu_basecore/MdePkg and you are using Efi prefix you are conveying that you are targeting UEFI spec.  If you were to remove Efi prefix and put it in MdeModulePkg you could target edkII or mu depending on prefix.

rayrobles replied:
I went ahead and moved the Media Sanitize protocol from MdePkg to MdeModulePkg (removing the "EFI_" prefix). Please look over my changes to ensure I did this properly. Thanks.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+  EFI_BLOCK_IO_MEDIA           *Media;
+
+  ///
+  /// Sanitize action shall be specific to the sanitize operation, such as
```
having a little trouble figuring out what the SanitizeAction field is for.  

Does this mean what are possible or supported values for the PurgeAction when a caller uses MediaPurge function?

rayrobles replied:
Correct. Per the NIST 800-88r1 spec., Section 5, Table 5-1, there are multiple methods to perform a purge...

"Some methods of purging (which vary by media and must be applied with considerations described further throughout this document) include overwrite, block erase, and Cryptographic Erase, through the use of dedicated, standardized device sanitize commands that apply media-specific techniques to bypass the abstraction inherent in typical read and write commands."

The intent of the SanitizeAction field is to capture the intended method of purge by the caller, and map to the transport specific method of carrying out the purge actions. In the case of NVMe, it is via the admin command Sanitize, which also contains a "SANIACTION" field that directly maps to overwrite, block erase, or crypto erase.

rayrobles replied:
Converted the SanitizeAction field in the protocol to SanitizeCapabilities. Use the check of sani capabilities when mapping the purge action to NVMe santize actions. Required update to the Industry Standard file Nvme.h in MdePkg for sanitize fields/structures. Please re-review at your convenience.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+  EfiBlockSanitizeCryptoErase   = 0x00000008,
+  EfiBlockSanitizeResetRequired = 0x00000010,
+  EfiBlockSanitizeNoDeallocate  = 0x00000020
+} EFI_BLOCK_SANITIZE_ACTION;
```
Not sure this should be an enum if you are going to make it unique bits.  Happy to be wrong but I think the convention would be to use defines if it is a bitfield where multiple bits can be set.  If it will only be one of these at any given time, then an enum is great.

rayrobles replied:
Agreed. I started down the typical enum path, but later decided that multiple bits could be set in the flag. In addition, I wanted to allow for growth/expansion of additional bits. 

I will convert the enum to #defines.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+  IN     EFI_MEDIA_SANITIZE_PROTOCOL   *This,
+  IN     UINT32                        MediaId,
+  IN     UINT32                        PurgeAction,
+  IN     UINT32                        OverwritePattern
```
Is the overwrite pattern only used when PurgeAction is EfiBlockSanitizeOverwrite

spbrogan replied:
Should the overwrite pattern be locked to 32bit value?  Does this give enough flexibility for future devices or other device types besides NVME?

rayrobles replied:
> Is the overwrite pattern only used when PurgeAction is EfiBlockSanitizeOverwrite

Correct. For crypto erase, the crypto keys are deleted only. For block erase, the data is "indeterminate" after the erase. Overwrite is the only action that specifies a pattern that shall be overwritten over the media.

> Should the overwrite pattern be locked to 32bit value? Does this give enough flexibility for future devices or other device types besides NVME?

Good question. Sanitize at a transport level for mass storage devices is typically defined by a pattern (vs. a buffer of data like with overwrite). The size of the pattern varies depending on the transport protocol. SATA/ATA, SAS, and NVMe all define the overwrite pattern as 4 bytes (or 32 bits).

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+/// EFI_MEDIA_SANITIZE_ACTION
+///
+typedef enum {
+  EfiBlockSanitizeNoAction      = 0x00000001,
```
more details on what each of these mean should be in this file

rayrobles replied:
Agreed. I will add comments describing each of the actions... and which fields can be overlapped.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+  // Defines whether the signaled event encountered an error.
+  //
+  EFI_STATUS    TransactionStatus;
+} EFI_MEDIA_SANITIZE_TOKEN;
```
I don't ever see this being used and I don't see an Async pattern supported here.  Did i miss something?

rayrobles replied:
I was originally planning on having a sync and async path, but after talking with OEMs, realized that a blocking/sync operation was sufficient (and actually desired). 

I will remove the structure as it's not needed.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+/** @file
+  This file defines the EFI Media Sanitize Protocol.
+
+  Copyright (c) 2016, Intel Corporation. All rights reserved.
```
Is this really an Intel authored protocol?

rayrobles replied:
No, I developed the protocol. I will update the copyright info. 

Is there a preferred style/format?

spbrogan replied:
look for other microsoft copyrights.  You should see a consistent style with no year.   That is what CELA has requested.

rayrobles replied:
Understood. I will make the updates based on existing Microsoft copyrights in other files.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,160 @@
+  @retval EFI_DEVICE_ERROR      The device reported an error while performing the write.
+  @retval EFI_NO_MEDIA          There is no media in the device.
+  @retval EFI_MEDIA_CHNAGED     The MediaId does not matched the current device.
+  @retval EFI_BAD_BUFFER_SIZE   The Buffer was not a multiple of the block size of the device.
```
this doesn't make sense

rayrobles replied:
Agreed. I will clean up the function header comments in all files updated/added.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,160 @@
+  @retval EFI_NO_MEDIA          There is no media in the device.
+  @retval EFI_MEDIA_CHNAGED     The MediaId does not matched the current device.
+  @retval EFI_BAD_BUFFER_SIZE   The Buffer was not a multiple of the block size of the device.
+  @retval EFI_INVALID_PARAMETER The write request contains LBAs that are not valid,
```
nor does this

rayrobles replied:
Agreed. Same action here as above.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,473 @@
+  } else if (PurgeAction & EfiBlockSanitizeCryptoErase) {
+    SanitizeAction = SANITIZE_ACTION_CRYPTO_ERASE;
+  } else {
+    SanitizeAction = 0;
```
does it make sense to do more error checking here?  Invalid values or unsupported values.  

Also what does action = 0 mean?  What do you do?

rayrobles replied:
Yes, it does. The call to NvmExpressSanitize() will perform other checks internally at the NVMe layer... so I was trying to avoid doing too much NVMe-specific checking in this MediaPurge function. But I do see the value of adding more "official" checking in this function to demonstrate what is and isn't acceptable at this protocol layer. I will update this mapping to include checks for invalid and unsupported values.

A purge action of 0 means perform no sanitize action (i.e a Sanitize no-op). My idea was to let the end storage device fail the transport-specific "Sanac" invalid field. I want to stay away from the transport-specific type checking that should be performed by the end device. Do you see a case to where I should put the check here instead?

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,473 @@
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+
+  ZeroMem (&CommandPacket, sizeof (EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET));
```
I prefer to see sizeof(Command) to keep type errors minimized.  There are pros and cons to this, so I would defer to the "style" from the nvme driver.  You can disregard this comment if the style is to use the type.

rayrobles replied:
Understood. I was trying to keep parity with the NVMe driver "style". Below is a sample of the ZeroMem in NvmeIdentifyController() inside NvmExpressHci.c...

  ZeroMem (&CommandPacket, sizeof (EFI_NVM_EXPRESS_PASS_THRU_COMMAND_PACKET));
  ZeroMem (&Command, sizeof (EFI_NVM_EXPRESS_COMMAND));
  ZeroMem (&Completion, sizeof (EFI_NVM_EXPRESS_COMPLETION));

I don't have a preference either way. I'm fine implementing either way. Just let me know which direction to follow.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,473 @@
+  Command.Nsid                 = NamespaceId;
+
+  DEBUG ((
+    DEBUG_ERROR,
```
Is this really an error?

rayrobles replied:
No, definitely not an error. I removed a check, but kept the debug statement. I will update to DEBUG_VERBOSE or DEBUG_INFO.

spbrogan commented on MdePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,147 @@
+/** @file
```
might want to add tpl restrictions as the implementation of nvme definitely has assumptions it is being called from application or callback level

rayrobles replied:
Agreed. I will update.

rayrobles replied:
Could you please clarify your comment? Are you looking for comments stating the functions must be called from a certain TPL? Or are you looking for run time checks to ensure TPL is at callback or application?

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressDxe.inf:
```diff
@@ -75,6 +76,7 @@
   gEfiStorageSecurityCommandProtocolGuid      ## BY_START
   gEfiDriverSupportedEfiVersionProtocolGuid   ## PRODUCES
   gEfiResetNotificationProtocolGuid           ## CONSUMES
+  gEfiMediaSanitizeProtocolGuid               ## CONSUMES
```
isn't this produces?

rayrobles replied:
Not sure actually. I view this protocol similar to Block Io, Disk Info, Storage Security, etc.

If "produces" is correct, I can change.

rayrobles replied:
Understood. Will make "produces".

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -166,6 +166,15 @@ EnumerateNvmeDevNamespace (
+    //
+    Device->MediaSanitize.Revision       = EFI_MEDIA_SANITIZE_PROTOCOL_REVISION;
+    Device->MediaSanitize.Media          = &Device->Media;
+    Device->MediaSanitize.SanitizeAction = 0;
```
why zero?

rayrobles replied:
Zero indicates to action to take for sanitize. Similar to setting a status to 0. 

Do you have a recommendation on another default beginning value and protocol install time?

rayrobles replied:
Resolved by adding #define for No Action (zero). Will also convert Sanitize Action to Sanitize Capabilities.

spbrogan reviewed (changes requested):
There are few broad patterns, headers, comments, copyrights that need to be cleaned up.  A few questions about plans regarding industry specs and the usage of Efi that need answers.

rayrobles committed:
Subset of modifications from code review comments - Part 1

rayrobles committed:
Subset of modifications from code review comments - Part 1

rayrobles committed:
Subset of modifications from code review comments - Part 2

rayrobles committed:
Subset of modifications from code review comments - Part 2

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles committed:
Subset of modifications from code review comments - Part 3

rayrobles committed:
Subset of modifications from code review comments - Part 3

rayrobles committed:
Subset of modifications from code review comments - Part 4

rayrobles committed:
Subset of modifications from code review comments - Part 4

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -166,6 +167,15 @@ EnumerateNvmeDevNamespace (
+    Device->MediaSanitize.Media      = &Device->Media;
+    Device->MediaSanitize.MediaClear = NvmExpressMediaClear;
+    Device->MediaSanitize.MediaPurge = NvmExpressMediaPurge;
+    CopyMem (&(Device->MediaSanitize.SanitizeCapabilities), &(Device->Controller->ControllerData->Sanicap), sizeof (NVME_SANICAP));
```
this should be sizeof(Device->MediaSanitize.SanitizeCapabilities)  or sizeof(UINT32)  because you want to make sure you never overflow the destination.  

That said a StaticAssert might also be good as you want to catch the case when they are different and break the build.

rayrobles replied:
Agreed... and updated to user sizeof(Device->MediaSanitize.SanitizeCapabilities). Also added an assert in the line before to check that both var sizes are equal.

rayrobles committed:
Updated CopyMem() to use sizeof(Device->MediaSantize.SanitizeCapabilities); also added a static assert to check theses sizes match

rayrobles committed:
Updated CopyMem() to use sizeof(Device->MediaSantize.SanitizeCapabilities); also added a static assert to check theses sizes match

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles committed:
Added TPL restriction comment to function headers; removed saving of old TPL from Media Clear and Purge functions

rayrobles committed:
Added TPL restriction comment to function headers; removed saving of old TPL from Media Clear and Purge functions

rayrobles committed:
Added BLOCK_MEDIA_FORMAT() function to Media Sanitize protocol; and mapping to NvmExpressMediaFormat()

rayrobles committed:
Added BLOCK_MEDIA_FORMAT() function to Media Sanitize protocol; and mapping to NvmExpressMediaFormat()

rayrobles commented:
Added BLOCK_MEDIA_FORMAT() function to Media Sanitize protocol. Also, added mapping to NvmExpressMediaFormat() in NVMe DXE driver. Updated the work doc attached to include BLOCK_MEDIA_FORMAT.

rayrobles commented:
Having issues uploading the modified Word Doc. So here is the new function...

1.3 MEDIA_SANITIZE_PROTOCOL.MediaFormat()

**Summary**

Format Media utilizes native format operations to modify sector/LBA size.  Secure erase actions are used to define how latent user data is erased.

**Prototype**

```
Typedef
EFI_STATUS
(EFIAPI *BLOCK_MEDIA_FORMAT)(
    IN MEDIA_SANITIZE_PROTOCOL	  *This,
    IN UINT32                                	  MediaId,
    IN UINT32 			 	          LbaSize, 
    IN UINT32 				          SecureEraseAction
);
```

**Parameters**

This		                 Indicates a pointer to the calling context.
MediaId		         The media ID that the clear operation is for.
LbaSize	                 This field specifies the new size of the sector/LBA on media. This size is defined in terms of a power of two 
                                 (2^n). Note this new size only applies to the media associated with the Media Id for the Block Io device. To 
                                 apply the new sector/LBA size across an entire physical device, the caller must invoke the format function 
                                 for all Media Id’s on the physical device. 
SecureEraseAction 	This field specifies whether a secure erase should be performed as part of the format operation and the 
                                 type of secure erase action. Secure erase actions supported include:

                                     - No secure erase action requested
                                     - User Data Erase
                                     - Cryptographic Erase

jeffglaum commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,480 @@
+  // Current supported LBA format size in Identify Namespace LBA Format Table, indexed by
+  // FLBAS (bits 3:0).
+  //
+  LbaFormat           = (!LbaFormat ? Device->NamespaceData.Flbas : Flbas);
```
What's the plan for supporting changing LBA size (this is always taking the current size)?

rayrobles replied:
Good catch! This is a bug. Line 98 was meant to be checking if the Flbas parameter passed in was NULL or not... and only changing the LBA size if the param is not NULL. Will resolve line to look like...

LbaFormat = (!Flbas ? Device->NamespaceData.Flbas : Flbas);

jeffglaum commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,586 @@
+      }
+
+      if ((FnvmSes == SES_USER_DATA_ERASE) || (FnvmSes == SES_CRYPTO_ERASE)) {
+        Status = NvmExpressFormatNvm (
```
In this case the specified overwrite pattern won't be applied following the format, correct?  Yes, the function will presumably return success.  Is this reasonable convention or is there facility for letting the caller know of the adaptation?

rayrobles replied:
Yes, this is a reasonable expectation for the NVMe admin command FormatNVM. Per NVMe specification, revision 1.4c, Section 5.23, Figure 332 (Format NVM - Command Dword 10):

For bits 11:09, Secure Erase Settings (SES) valid values are...
  - 000b: No secure erase operation requested
  - 001b: User Data Erase
  - 010b: Cryptographic Erase
  - 011b to 111b: Reserved

Format NVM does not support an "Overwrite" erase operation. Therefore, if Sanitize fails (with the overwrite sanitize action), the code will only retry with Format NVM iff the erase operation is User Data erase or Cryptoerase.

The return status will actually still have the failure code (opcode not supported) from the failed Sanitize command in this case. The check at line 289 will only send the Format NVM if the erase action is User Data or Crypto. Therefore, the failed "Overwrite" operation, and its associated failure status, will maintain state and be returned back to the caller.

jeffglaum commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,586 @@
+  failed sanitize operation. The sanitize operation types that may be supported are Block
+  Erase, Crypto Erase, and Overwrite.
+
+  All sanitize operations are processed in the background (i.e., completion of the Sanitize
```
Are you intending this to be a blocking call (other references to the function seem to suggest as much)?  The pass-through call below is only issuance of the command, not confirmation it's completed, correct?  Maybe I'm missing it but I don't see the wait for completion below.

rayrobles replied:
Correct. For this initial revision, both NVMe admin commands added (Sanitize and FormatNVM) are implemented to be blocking calls. Per direct feedback from Dell and Lenovo (and QC also), the preference is that these "erase" operations be blocking/synchronous calls.

**Dell Feedback:** "It is expected that once the data wipe is launched, the drives will complete the data wipe, even if there is a power event. This does lend itself to be somewhat asynchronous. However, it is expected that the operation completes before booting is allowed."

**Lenovo Feedback:** "We don’t see any necessities for asynchronous even if it takes longer time"

Note that the implementation does allow for asynchronous completions to be supported. The NVMe passthru mechanism allows for asynchronous completions (just requires an additional event structure to be used and passed).

jeffglaum commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,586 @@
+
+  NOTE: The caller shall send buffer of one sector/LBA size with overwrite data.
+  NOTE: This operation is a blocking call.
+  NOTE: This function must be called from TPL aaplication or callback.
```
NIT: Does it make sense to check/assert this condition at entry?

rayrobles replied:
Good question. Sean Brogan had similar feedback as well. When I asked him whether he wanted me to add a runtime check or just a comment in the code, he stated a comment was fine. Again, we can always add runtime checks at a later time if we feel the need.

makubacki replied:
I think it makes sense to check it. There are places in the code that violate TPL calling restrictions but cannot be updated (easily) because so much other code now depends on it.

jeffglaum commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,586 @@
+
+ **/
+EFI_STATUS
+NvmExpressFormatNvm (
```
Is the intent that this and NvmExpressSanitize are private and only accessed through NvmExpressMedia* functions?  Not sure if there's convention for keyword tagging or splitting into a private-side source file for clarity if that's the case but might be good to at least clarify in the comment.

rayrobles replied:
Yes, as implemented in this PR, NvmExpressFormatNVM and NvmExprssSanitize shall only be called from within the NVMe UEFI driver. Both of these functions do not have the "EFIAPI" define associated, so therefore will not be exported outside the NVMe DXE environment. 

Note that the three new Media Sanitize Protocol Functions all have the "EFIAPI" defined. This allows them to be exported externally and be visible to other UEFI DXE environments.

EFI_STATUS
EFIAPI
NvmExpressMediaClear (
  IN MEDIA_SANITIZE_PROTOCOL  *This,
  IN UINT32                   MediaId,
  IN UINT32                   PassCount,
  IN VOID                     *SectorOwBuffer
  );

EFI_STATUS
EFIAPI
NvmExpressMediaPurge (
  IN MEDIA_SANITIZE_PROTOCOL  *This,
  IN UINT32                   MediaId,
  IN UINT32                   PurgeAction,
  IN UINT32                   OverwritePattern
  );

EFI_STATUS
EFIAPI
NvmExpressMediaFormat (
  IN MEDIA_SANITIZE_PROTOCOL  *This,
  IN UINT32                   MediaId,
  IN UINT32                   LbaSize,
  IN UINT32                   SecureEraseAction
  );

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles commented:
[Media Sanitize Protocol.docx](https://github.com/microsoft/mu_basecore/files/8683205/Media.Sanitize.Protocol.docx)

rayrobles committed:
Resolve bug from code review (Jeff Glaum) on checking correct Flbas parameter in NvmExpressFormatNvm

rayrobles committed:
Resolve bug from code review (Jeff Glaum) on checking correct Flbas parameter in NvmExpressFormatNvm

rayrobles committed:
Merge branch 'release/202202' of https://github.com/rayrobles/efi_media_sanitize_protocol into release/202202

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles commented:
> There are few broad patterns, headers, comments, copyrights that need to be cleaned up. A few questions about plans regarding industry specs and the usage of Efi that need answers.

Changes have been completed from reviews for both @spbrogan and @jeffglaum

rayrobles committed:
Fix outdated comments

rayrobles committed:
Fix outdated comments

rayrobles committed:
Merge branch 'release/202202' of https://github.com/rayrobles/efi_media_sanitize_protocol into release/202202

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles committed:
Port of Media Sanitize unit tests to GitHub repo (from ADO)

rayrobles committed:
Port over media sanitize unit tests from ADO repo

rayrobles commented:
@spbrogan I've checked in my unit tests for this PR. Please let me know if I need to do anything else in order to get this PR approved.

Thanks again for all your help!

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTestHost.inf:
```diff
@@ -0,0 +1,41 @@
+## @file
+# Unit tests of the DxeResetSystemLib instance of the ResetSystemLib class
```
This comment needs to be updated for MediaSanitizeUnitTestHost

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+#include "../NvmExpressHci.h"
+#include "../../../Include/Protocol/MediaSanitize.h"
+
+// #pragma warning ( disable : 4100 )
```
What specifically needed this warning disabled?

rayrobles replied:
I had originally the NvmeRead() and NvmeWrite() functions as stubs (not doing anything inside). So I added the #pragma to suppress the unreferenced parameter warning. I've since fixed these two functions to address the parameters so I could remove the #pragma. I simply forgot to remove line #28. I will delete the line.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+                                 allocated and built. Caller must set the NamespaceId to zero if the
+                                 device path node will contain a valid UUID.
+  @param[in] SanitizeAction      Sanitize action
+  @param[in] NoDeallocAfterSani  No deallocate after sanitize option
```
Parameter name should be `NoDeallocAfterSanitize`.

rayrobles replied:
Agreed. Not sure what happened to truncate it. I will update line #70.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+                                 device path node will contain a valid UUID.
+  @param[in] SanitizeAction      Sanitize action
+  @param[in] NoDeallocAfterSani  No deallocate after sanitize option
+  @param[in] OverwritePatter     Pattern to overwrite old user data
```
Parameter name should be `OverwritePattern`.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+  storage devices as well as to manage such devices in the EFI boot services
+  environment.
+
+  @param  This             Indicates a pointer to the calling context.
+  @param  MediaId          The media ID that the write request is for.
+  @param  PassCount        The number of passes to write over media.
+  @param  SectorOwBuffer   A pointer to the overwrite buffer.
```
These paramter descriptions are missing the `[IN]` tag.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+  NOTE: This operation is a blocking call.
+  NOTE: This function must be called from TPL aaplication or callback.
+
+  @param  This             Indicates a pointer to the calling context.
+  @param  MediaId          The media ID that the write request is for.
+  @param  PurgeAction      The purage action (overwrite, crypto erase, block erase).
+  @param  OverwritePattern 32-bit pattern to overwrite on media (for overwrite).
+
```
These paramter descriptions are missing the [IN] tag.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+
+  NOTE: The caller shall send buffer of one sector/LBA size with overwrite data.
+  NOTE: This operation is a blocking call.
+  NOTE: This function must be called from TPL aaplication or callback.
```
Typo: "aaplication" -> "application"

Consider using `TPL_APPLICATION` and `TPL_CALLBACK` to catch searches for those.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+  Format Media utilizes native format operations to modify sector/LBA size.
+  Secure erase actions are used to define how latent user data is erased.
+
+  NOTE: This function must be called from TPL aaplication or callback.
```
Same as L96.

Typo: "aaplication" -> "application"

Consider using TPL_APPLICATION and TPL_CALLBACK to catch searches for those.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.h:
```diff
@@ -0,0 +1,191 @@
+                                 - 010b: Cryptographic Erase
+                                 - 011b to 111b: Reserved
+
+  @retval EFI_SUCCESS             The media format request comopleted successfully on the device.
```
Type: "comopleted" -> "completed"

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressDxe.inf:
```diff
@@ -40,6 +40,7 @@
   NvmExpressHci.c
   NvmExpressHci.h
   NvmExpressPassthru.c
+  NvmExpressMediaSanitize.c
```
Please add the header file to `[Sources]` as well. It will usually show up as a warning if missing.

rayrobles replied:
Will add the header file on next line.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,574 @@
+    // cached copies of fields related to block size.
+    //
+    if (Flbas) {
+      NVME_ADMIN_NAMESPACE_DATA  *NewNamespaceData;
+      UINT32                     Lbads;
+      UINT32                     NewFlbas;
+      UINT32                     LbaFmtIdx;
```
We don't normally declare variables at block scope within a function. @spbrogan, I'd assume we want to continue enforcing that?

rayrobles replied:
I can move these variable declarations to the top of the function is that is the BKM for Mu. Will update. Thanks.

rayrobles replied:
Moved all variable declarations to top of their respective functions.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressMediaSanitize.c:
```diff
@@ -0,0 +1,574 @@
+      if (NewNamespaceData == NULL) {
+        Status = EFI_OUT_OF_RESOURCES;
+      } else {
+        DEBUG ((DEBUG_VERBOSE, "NvmExpressFormatNvm: Format NVM SUCCEEDED; updating namesapce data\n"));
```
Minor but consider using the function name. It helps when people copy/paste for new messages and forget to update and keep the meessages accurate if the function name is actually updated.

```c
DEBUG ((DEBUG_VERBOSE, "%a: Format NVM SUCCEEDED; updating namespace data\n", __FUNCTION__));
```

There was also a typo: "namesapce" -> "namespace"

rayrobles replied:
Good point. Will updates.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -166,6 +167,26 @@ EnumerateNvmeDevNamespace (
+      sizeof (Device->Controller->ControllerData->Sanicap)
+      );
+
+    CopyMem (
+      &(Device->MediaSanitize.SanitizeCapabilities),
+      &(Device->Controller->ControllerData->Sanicap),
+      sizeof (Device->MediaSanitize.SanitizeCapabilities)
+      );
```
Should the `CopyMem()` still happen if the `ASSERT` is TRUE? I don't think there's a risk of overflowing the destination buffer, but I just wanted to check if that makes sense given how often these fields may be managed.

kuqin12 replied:
Given this assert is comparing the field width instead of dynamic size from data, should this be using `STATIC_ASSERT`?

rayrobles replied:
Good question. The original intent of the of the assert check is due to the fact that I'm comparing a 32 bit structure to a UINT32 (SanitizeCapabilities in the MEDIA_SANITIZE_PROTOCOL). Since the "sanicap" structure is an NVM Express construct, I was trying to leave out that level of detail in the MEDIA_SANITIZE_PROTOCOL header file. Therefore, I just made the field a UINT32. It's really just a sanity check to see if anyone decides to change the size of the "sanicap" structure for NVM Express, the correlating change in the agnostic MEDIA_SANITIZE_PROTOCOL must also be changed. If you feel strongly, I can add a run time check instead of an assert. Just let me know.

makubacki replied:
Thanks for the explanation. I'm okay with leaving it as-is.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTestHost.inf:
```diff
@@ -0,0 +1,41 @@
+#  VALID_ARCHITECTURES           = IA32 X64
+#
+
+[Sources]
```
Please add the header files here as well.

rayrobles replied:
Agreed. Will update.

makubacki commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+/** @file -- MediaSanitizeUnitTest.c
```
Throughout the file there seems to be some commented code that could be cleaned up.

I recommend deleting it if it's not needed at all. If there's intermediate changes for a future change, delete it from here and keep in a local branch.

rayrobles replied:
Agreed. Will remove commented out code.

makubacki commented on MdePkg/Include/IndustryStandard/Nvme.h:
```diff
@@ -663,6 +708,9 @@ typedef struct {
   UINT32    Pi    : 3;        /* Protection Information */
   UINT32    Pil   : 1;        /* Protection Information Location */
   UINT32    Ses   : 3;        /* Secure Erase Settings */
+  #define SES_NO_SECURE_ERASE  0x0
+  #define SES_USER_DATA_ERASE  0x1
+  #define SES_CRYPTO_ERASE     0x2
```
I suggest moving these outside of the structure definition given our usual style.

rayrobles replied:
I was a bit unsure here which style to follow. The nvme.h file has several instances where #defines are declared right inside the structure (see below examples). So I was trying to stick to precedent. However, if Mu uses a different convention, I'm fine moving the #defines outside the structure definition. 

typedef struct {
  //
  // CDW 10
  //
  UINT32    Lid   : 8;        /* Log Page Identifier */
  #define LID_ERROR_INFO    0x1
  #define LID_SMART_INFO    0x2
  #define LID_FW_SLOT_INFO  0x3
  #define LID_BP_INFO       0x15
  UINT32    Rsvd1 : 8;
  UINT32    Numd  : 12;       /* Number of Dwords */
  UINT32    Rsvd2 : 4;        /* Reserved as of Nvm Express 1.1 Spec */
} NVME_ADMIN_GET_LOG_PAGE;

makubacki replied:
I saw that after I left the comment. I think those are violating the normal convention as well. Since it's already done that way here elsewhere, I am fine with it being left as-is.

rayrobles replied:
No worries. I went ahead and pulled the ones I added out of the structures.

makubacki commented on MdePkg/Include/IndustryStandard/Nvme.h:
```diff
@@ -698,6 +746,30 @@ typedef struct {
+  // CDW 10
+  //
+  UINT32    Sanac : 3;        /* Sanitize Action */
+  #define SANITIZE_ACTION_NO_ACTION          0x0
+  #define SANITIZE_ACTION_EXIT_FAILURE_MODE  0x1
+  #define SANITIZE_ACTION_BLOCK_ERASE        0x2
+  #define SANITIZE_ACTION_OVERWRITE          0x3
+  #define SANITIZE_ACTION_CRYPTO_ERASE       0x4
```
Suggest same as above to move macros outside the structure definition.

If these represent bits, we typically use `BIT01`, `BIT1`, etc. from Base.h.

rayrobles replied:
Same response as above. I'll go ahead and move the #defines outside the structure definition. 

In this case, the #define values do not map to bits, but values as defined by the NVMe specification, Section 5.24 - Sanitize command, Figure 334 - Sanitize Command Dword 10... Bits 02:00 (Sanitize Action)...

Value Description
000b Reserved
001b Exit Failure Mode
010b Start a Block Erase sanitize operation
011b Start an Overwrite sanitize operation
100b Start a Crypto Erase sanitize operation
101b to 111b Reserved

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+#include "../NvmExpressBlockIo.h"
+#include "../NvmExpressMediaSanitize.h"
+#include "../NvmExpressHci.h"
+#include "../../../Include/Protocol/MediaSanitize.h"
```
you should be able to use `<Protocol/MediaSanitize.h>` as this is a public header file.

rayrobles replied:
Agreed. Will update.

rayrobles replied:
UPDATE: @spbrogan I could not build -NOOPT with your suggestion here. I needed to provide the relative path. The host based unit test could not see the protocol library since I was just creating a small executable with NVMe and unit test source.

spbrogan replied:
that doesn't make sense to me.  I'll look into this one.

spbrogan replied:
I was able to build and pass the tests with this change
``` c
//sean#include "../../../../Include/Protocol/MediaSanitize.h"
#include <Protocol/MediaSanitize.h>
```

rayrobles replied:
Thanks for looking into this one @spbrogan. I figured out what the issue was for me. 

**When I added the #include line before the other relative NVM Express include paths** (like below), I got the compile error.

#include <Uefi.h>
#include <Library/BaseLib.h>
#include <Library/BaseMemoryLib.h>
#include <Library/UefiLib.h>
#include <Library/DebugLib.h>
#include <Library/UnitTestLib.h>
#include <Library/UefiBootServicesTableLib.h>
#include <Library/UefiApplicationEntryPoint.h>
#include <Library/UefiRuntimeServicesTableLib.h>
#include <Library/UefiBootManagerLib.h>
#include <Protocol/NvmExpressPassthru.h>
**#include <Protocol/MediaSanitize.h>**

#include "../NvmExpress.h"
#include "../NvmExpressBlockIo.h"
#include "../NvmExpressMediaSanitize.h"
#include "../NvmExpressHci.h"

However, when I add the MediaSanitize.h header include **after the local, relative NVM Express header includes** (see below), I got a successful compile and run.

#include <Uefi.h>
#include <Library/BaseLib.h>
#include <Library/BaseMemoryLib.h>
#include <Library/UefiLib.h>
#include <Library/DebugLib.h>
#include <Library/UnitTestLib.h>
#include <Library/UefiBootServicesTableLib.h>
#include <Library/UefiApplicationEntryPoint.h>
#include <Library/UefiRuntimeServicesTableLib.h>
#include <Library/UefiBootManagerLib.h>
#include <Protocol/NvmExpressPassthru.h>

#include "../NvmExpress.h"
#include "../NvmExpressBlockIo.h"
#include "../NvmExpressMediaSanitize.h"
#include "../NvmExpressHci.h"
**#include <Protocol/MediaSanitize.h>**

I will update the file to use the public include. Thanks!

spbrogan replied:
after more review you have fixed it correctly by finding the real dependency (protocol/blockio.h).  Might be helpful to those trying to use the protocol if a comment was added in the header file saying the definition depends on blockio.h.  

if you have no other changes to make...please disregard and just resolve.  :)

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  EFI_STATUS                  Status;
+  NVME_CQ                     *Cqe;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeDeviceUnitTestPassthru: Enter\n"));
```
to make more portable code we often use %a format string for the function name and then __FUNCTION__ as a parameter.  Avoids having to keep correct.

rayrobles replied:
Makes sense. Will update.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  EFI_STATUS                  Status;
+  NVME_CQ                     *Cqe;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeDeviceUnitTestPassthru: Enter\n"));
```
Are the enter/exit messages useful now that the test is developed?

rayrobles replied:
Good question. No, they are not. It was useful as I was developing the unit tests and getting an understanding of the flow of how the host application was invoked. I will remove the enter/exit debug print statements.

spbrogan replied:
adding to it.  Can you help explain what the "[         ]" part of the debug message is for.    Also there is a UT_LOG_* macro that will do extra logging for the unit tests and get it added to the report.

rayrobles replied:
The extra "[  ]" was to make the output look similar to the other UT outputs. I can convert all the debug prints to use the UT LOG macro if that is preferred. I was unaware of the logging macro when I first wrote the file... this would be preferred for me as well.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiApplicationEntryPoint.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Uefi/UefiBaseType.h>
```
i don't think basetype is needed given you included uefi.h at the top.

rayrobles replied:
Agreed. Will remove.

kuqin12 commented on MdeModulePkg/Include/Protocol/MediaSanitize.h:
```diff
@@ -0,0 +1,173 @@
+#define PURGE_ACTION_BLOCK_ERASE                       0x00000002 // Erase Blocks with indeterminate pattern
+#define PURGE_ACTION_CRYPTO_ERASE                      0x00000004 // Delete encryption keys only
+#define PURGE_ACTION_RESET_REQUIRED                    0x00000008 // Reset required after purge
+#define PURGE_ACTION_NO_DEALLOCATE                     0x00000010 // Do no deallocate (trim) flash medai after sanitize
```
nit: should it be below?
```suggestion
#define PURGE_ACTION_NO_DEALLOCATE                     0x00000010 // Do not deallocate (trim) flash medai after sanitize

```

spbrogan replied:
also fix spelling for media

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+#include "../NvmExpressBlockIo.h"
+#include "../NvmExpressMediaSanitize.h"
+#include "../NvmExpressHci.h"
+#include "../../../Include/Protocol/MediaSanitize.h"
```
Should this be a normal inclusion `<Protocol/MediaSanitize.h>`?

rayrobles replied:
Agreed. I was having an issue with the public include, but resolved the issue after moving the public include after the local NVMe header includes. Will update.

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  NVME_DEVICE_PRIVATE_DATA  *NvmeDevice    = NULL;
+  EFI_STATUS                Status         = EFI_SUCCESS;
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeSanitizenitTest: Starting Unit Test\n"));
```
```suggestion
  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeSanitizeUnitTest: Starting Unit Test\n"));

```

Also, it might be good to switch to use below to save the duplicated function name spell-outs.
```suggestion
  DEBUG ((DEBUG_VERBOSE, "%a: Starting Unit Test\n", __FUNCTION__));

```

rayrobles replied:
Agree. Will update to use __FUNCTION__.

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  UT_ASSERT_NOT_EFI_ERROR (Status);
+ #endif
+
+ #if 0
```
Is this right? Will we leave this out in the final check-in?

rayrobles replied:
This was originally for my own debug purposes. I have removed all debug #defines.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  DEBUG ((DEBUG_VERBOSE, "[          ] NvmeDeviceUnitTestPassthru: Enter\n"));
+
+  UT_ASSERT_NOT_NULL (This);
+  UT_ASSERT_NOT_NULL (Packet);
```
this isn't a test function so not sure you should really use UT asserts in it. UT asserts are for test functions and the macros use the UT logging and error codes.  This function returns EFI_STATUS codes.

rayrobles replied:
Good point. I'll remove the UT_ASSERT...() calls.

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+
+  DEBUG ((DEBUG_VERBOSE, "[          ] MediaSanitizePurgeUnitTest: Starting Unit Test\n"));
+
+  UnitTestStatus = NvmeCreateDeviceInstance (&NvmeDevice);
```
Please ignore if I am not understanding it correctly, but I feel this function and `NvmeDestroyDeviceInstance` suits the purpose of `UNIT_TEST_PREREQUISITE` and `UNIT_TEST_CLEANUP`? It is nothing functionally different, but just to leverage the framework to setup the environment.

rayrobles replied:
It could definitely be leveraged through the prerequisite and cleanup functions, but I was having issues with the memory allocation of the emulated NVMe device handling. Are you okay if I resolve this in another PR in the near future? I do agree this should be moved, I would just like to get the functional code part of this PR pushed first. @kuqin12

kuqin12 replied:
Yes, agreed. Let's improve the unit test separately.

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/UnitTest/MediaSanitizeUnitTest.c:
```diff
@@ -0,0 +1,1208 @@
+  }
+
+  Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO (This);
+  Status = NvmeUnitTestWrite (Device, Buffer, Lba, NumberOfBlocks);
```
Just curious, this function does not seem to do much inside, is that right?

If you would like to inspect the content inside, you can add cmocka functions inside so that you can set up expected parameter inspections.

rayrobles replied:
For now, that is correct. As more unit test cases get added there may be a need to examine the data itself that is passed (i.e. pattern recognition, buffer length, etc.). I would put this in the same category as the comment above. It can be addressed in a future PR to build more robustness to the unit tests.

kuqin12 replied:
Yes, agreed. Let's make that change in a separate PR.

spbrogan commented:
In case you can't see the build log which is failing on GCC.
[CI_BUILDLOG (9).txt](https://github.com/microsoft/mu_basecore/files/8765350/CI_BUILDLOG.9.txt)

rayrobles commented:
> In case you can't see the build log which is failing on GCC. [CI_BUILDLOG (9).txt](https://github.com/microsoft/mu_basecore/files/8765350/CI_BUILDLOG.9.txt)

Thanks @spbrogan! I can't seem to access the full build logs on GitHub. I can access the ADO build logs though. I see the error now and will fix it.

rayrobles committed:
First revision of chages for review feedback

rayrobles committed:
Second revision of updates from code review comments

rayrobles committed:
Resolve compile and uncrustify issues in latest set of updates for review comments

kuqin12 commented:
sorry, I was confused by the comment UI and only published the comments now..

rayrobles committed:
Final pass on PR code review comments

rayrobles commented:
@spbrogan, @kuqin12, @makubacki, @jeffglaum I believe I have resolved all the comments in this PR and have made all the requested updates. Please double check to make sure I didn't forget anything. Please let me know if there is anything else for this PR. Thanks!

rayrobles committed:
Fix gcc MdeModulePkg Debug compile issue for unsued, but set, variable

rayrobles committed:
Merge branch 'release/202202' into release/202202

rayrobles commented:
All reviewers have approved this PR. I've rebased from origin. Closing this PR for new MEDIA_SANITIZE_PROTOCOL that maps NIST purge, clear operations to NVMe native command support. Unit tests also included (host based).

rayrobles commented:
[Media Sanitize Protocol.docx](https://github.com/microsoft/mu_basecore/files/8788001/Media.Sanitize.Protocol.docx)
